keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -162,10 +163,81 @@ TODO: complete this list
-* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
-* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
-* Coordinate with sig-scalability
+The scale target for GA of custom resources are defined by the same [API call latency
+SLIs/SLOs as the Kuberetes native types](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md#api-call-latency-slisslos-details).
+
+The targets are defined by the below suggested maximum limits, which are organized the same way as the [Kubernetes native type thresholds](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds), with only one change:
+
+- Since custom resources can be arbitrarily large, we have broken down the limit by custom resource object size.
+
+**Custom Resources per Definition:**
+
+| Object size | Suggested Maximum Limit: scope=namespace (5s p99 SLO) | Suggested Maximum Limit: scope=cluster (30s p99 SLO) |
+| --- | --- | --- |
+| 10kb | 1500 | 10000 |
+| 25kb | 600 | 4000 |
+| 50kb | 300 | 2000 |
+
+To be consistent with Kubernetes native type SLOs, this SLO includes client serialization/deserialization.
+
+Since, in practice, custom resources scale farther without conversion webhooks
+within the SLI/SLOs (roughly 2x according to our scale tests), custom resource
+definition authors should be careful to adhere to these limits so that in the
+future a webhook converter may safely be added as part of a custom resource
+version upgrade.
+
+_Note: For custom resources of custom resource definitions using `scope: Namespaced`: the scope=namespace
+suggested maximum limit indicates how many custom resource objects may be in each namespace,
+and the scope=cluster suggested maximum limit indicates how many custom resource objects may
+be in the cluster total. For custom resources of custom resource definitions using `scope: Cluster`: only
+the scope=cluster suggested maximum limit applies._
+
+**Custom Resource Definitions:**
+
+| Suggested Maximum Limit: scope=cluster |
+| --- |
+| 500 |
+
+_Note: The Custom Resource Definition suggested maximum limit was selected not
+due to the above SLI/SLOs, but instead due to the latency OpenAPI publishing,
+which is a background process that occurs asychroniously each time a Custom
+Resource Definition schema is updated. For 500 Custom Resource Definitions it takes
+slightly over 35 seconds for a definition change to be visible via the OpenAPI
+spec endpoint._
+
+**Conversion Webhooks:**
+
+Given that the performance and scalability of conversion webhooks are the
+responsibility of their author, Custom resource scale targets are apply only for
+conversion webhooks that are within the follow latencies for the above suggested
+maximum limits.
+
+| scope | Expected conversion Webhook SLO: p99 latency |
+| --- | --- |
+| resource | 50ms | 
+| namespace | 1 seconds |
```

Comment:
I think we should be consistent in all those cases - since for resource we're saying 5% (for webhook), we should use 5% also here and for cluster. So 250ms and 1500ms respectively.

Reply:
That's reasonable. I'll update.

Reply:
Oh, actually, 5% is not enough for namespace and cluster, our benchmarks show we need 20%, I'll make this consistent by setting resource to 200ms.

Reply:
200ms for resource is too much - we won't be able to satisfy it on high enough load.

Reply:
Update: We're going to go with 50ms per out-of-band discussion. This is well over what a conversion webhook should need to serve the request (which is closer to 5ms based on our benchmarks), so there is quite a bit of extra headroom for request serving overhead.

We didn't want to go up to 100ms+ because (a) for day long p99s in 5k scale tests, etcd latency can be up to 600-700ms, so giving webhooks 200ms gets the overall request latency too high for our SLOs, and (b) we don't want to set an expectation with webhook conversion authoers that 100ms is acceptable. If they're anywhere near that number, their implementation is not performing well and is going to be a problem.